### ISSUE TITLE
Disantangle the Centerline VS Flowline class tree

### ISSUE THREAD

**Issue by fmaussion** [OGGM/oggm#561]
Certain attributes are only available at the Flowline level (e.g. `widths_m`) although it should not be.

The two classes have arisen as the code was growing, but it's time for a brainstorm about what these actually do.

**Comment by fmaussion** [comment]
Again, an organic development process lead to a hierarchy of classes which is suboptimal and complex to understand.

To date, I think only @matthiasdusch and myself have a kind of an overview of them (see also https://github.com/OGGM/oggm/pull/811).

It would be great to revisit this issue one day. Tagging as refactoring and quite important to my eyes.

**Comment by matthiasdusch** [comment]
I agree on the importance. Quite daunting at first glance. First thing would be to come up with a good concept how these classes should look like.

Do you think it would be feasible to build this around a single class and treat the different steps as states?

**Comment by fmaussion** [comment]
Let's do a brainstorm together next week, but:
- yes, removing inheritance would be much better
- possibly also having only one single class dealing with all kind of bed shapes would be good
- this is going to brake ALL pickles from previous OGGM versions. If we tackle this, then we should add a serialization to netCDF instead of relying on pickle to avoid future compatibility issues

> Quite daunting at first glance.

Yes, quite a lot of work and, again, very little reward on the short term (but big reward on the long term...)

**Comment by fmaussion** [comment]
Probably this is going to be one reason to work with a branch at first, then generate the new pre-processed dirs, then publish a new OGGM version. 

Maybe worth tackling together with the next generation of pre-processed dirs...
